lib/spot_price.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+require_relative './aws_helper'
+require 'datadog/statsd'
+
+class SpotPrice
+  def self.record_metrics()
+    new.record_metrics
+  end
+
+  def record_metrics()
+    # instance types we allow in Spot instance groups.
+    %w(c5.xlarge r5.xlarge m5.xlarge c6i.xlarge r6i.xlarge m6i.xlarge).each do |instance_type|
+      statsd.gauge "spot_price", get_spot_price(instance_type), {:tags => [instance_type]}
```

Comment:
opting to have 1 metric (spot_price) with instance type as tags (c5.xlarge, r5.xlarge, ...), rather than 1 metric per instance type (spot_price.c5.xlarge, spot_price.r5.xlarge, ...)

Reply:
a pain to test this locally, will see if it works when deployed.

Reply:
Why are you giving it an instance and a singleton method with the same name? I find that a bit confusing to read... Also with calling `new` inside the singleton method... Also the empty brakets without arguments are not necessary 🤔 The `get_spot_price` method also seems like it could be made `private`

Reply:
> giving it an instance and a singleton method with the same name...

Yes I take the credit/blame for this. My motivations were to offer a class-level helper method to minimize code in the rake task, but construct an instance to contain any memoized data such as configs, clients, connections, etc.

Memo-izing on classes is possible but I generally avoided that because it pollutes the global environment, is difficult to test, gets awkward with inheritance, creates race conditions in multi-threaded environments.... Some of these don't apply here, but just trying to explain all my scar tissue. 😅

Reply:
@joeyAghion while I understand your motivation for the weird anti pattern Ruby here, I think the reason why we are using Ruby is because it is collaborative in the sense that it can be very readable code and easy/quick to work on. Using very anti-pattern and unconventional Ruby does the opposite. It makes it hard to read and understand what exactly is going on and will take the next person longer to work on it. Sometimes saving lines of code is not the best way to go if that makes the code very implicit. I suggest to refactor the code into using an initialize method and calling instances of the class in the rake task and remove memoizing all along (pass the parameters into the initializer).